File: jambeth/jambeth-ioc/src/main/java/de/osthus/ambeth/util/CatchingRunnable.java
package de.osthus.ambeth.util;

import java.util.concurrent.CountDownLatch;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.RunnableFuture;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

import de.osthus.ambeth.ioc.threadlocal.IForkState;
import de.osthus.ambeth.ioc.threadlocal.IThreadLocalCleanupController;

public class CatchingRunnable implements RunnableFuture<Throwable>
{
	protected final IForkState forkState;

	protected final Runnable runnable;

	protected final CountDownLatch latch;

	protected final IParamHolder<Throwable> throwableHolder;

	protected final IThreadLocalCleanupController threadLocalCleanupController;

	public CatchingRunnable(IForkState forkState, Runnable runnable, CountDownLatch latch, IParamHolder<Throwable> throwableHolder,
			IThreadLocalCleanupController threadLocalCleanupController)
	{
		this.forkState = forkState;
		this.runnable = runnable;
		this.latch = latch;
		this.throwableHolder = throwableHolder;
		this.threadLocalCleanupController = threadLocalCleanupController;
	}

	@Override
	public Throwable get() throws InterruptedException, ExecutionException
	{
		latch.await();
		return throwableHolder.getValue();
	}

	@Override
	public Throwable get(long timeout, TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException
	{
		if (latch.await(timeout, unit))
		{
			return throwableHolder.getValue();
		}
		throw new TimeoutException("No result after " + timeout + " " + unit.name());
	}

	@Override
	public boolean isDone()
	{
		return latch.getCount() == 0;
	}

	@Override
	public boolean isCancelled()
	{
		return false;
	}

	@Override
	public boolean cancel(boolean mayInterruptIfRunning)
	{
		return false;
	}

	@Override
	public void run()
	{
		Thread currentThread = Thread.currentThread();
		String oldName = currentThread.getName();
		if (runnable instanceof INamedRunnable)
		{
			currentThread.setName(((INamedRunnable) runnable).getName());
		}
		try
		{
			try
			{
				if (forkState != null)
				{
					forkState.use(runnable);
				}
				else
				{
					runnable.run();
				}
			}
			catch (Throwable e)
			{
				throwableHolder.setValue(e);
			}
			finally
			{
				threadLocalCleanupController.cleanupThreadLocal();
				latch.countDown();
			}
		}
		finally
		{
			if (runnable instanceof INamedRunnable)
			{
				currentThread.setName(oldName);
			}
		}
	}
}

File: ambeth/Ambeth.IoC/ambeth/util/CatchingRunnable.cs
using De.Osthus.Ambeth.Config;
using De.Osthus.Ambeth.Ioc;
using De.Osthus.Ambeth.Ioc.Annotation;
using De.Osthus.Ambeth.Ioc.Threadlocal;
using De.Osthus.Ambeth.Model;
using De.Osthus.Ambeth.Threading;
using System;
using System.Threading;

namespace De.Osthus.Ambeth.Util
{
    public class CatchingRunnable : Runnable
    {
        protected readonly IForkState forkState;

	    protected readonly Runnable runnable;

	    protected readonly CountDownLatch latch;

	    protected readonly IParamHolder<Exception> throwableHolder;

        protected readonly IThreadLocalCleanupController threadLocalCleanupController;

        public CatchingRunnable(IForkState forkState, Runnable runnable, CountDownLatch latch, IParamHolder<Exception> throwableHolder,
			    IThreadLocalCleanupController threadLocalCleanupController)
	    {
		    this.forkState = forkState;
		    this.runnable = runnable;
		    this.latch = latch;
		    this.throwableHolder = throwableHolder;
		    this.threadLocalCleanupController = threadLocalCleanupController;
	    }
       
        public void Run()
        {
            Thread currentThread = Thread.CurrentThread;
            String oldName = currentThread.Name;
            if (runnable is INamedRunnable)
            {
                currentThread.Name = ((INamedRunnable)runnable).Name;
            }
            try
            {
                try
                {
                    if (forkState != null)
                    {
                        forkState.Use(runnable);
                    }
                    else
                    {
                        runnable.Run();
                    }
                }
                catch (Exception e)
                {
                    throwableHolder.Value = e;
                }
                finally
                {
                    threadLocalCleanupController.CleanupThreadLocal();
                    latch.CountDown();
                }
            }
            finally
            {
                currentThread.Name = oldName;
            }
        }
    }
}